### PR TITLE
feat(server): /v1/completions echo + logprobs + full legacy-API parity

### PR DESCRIPTION
**Pull request by tbraun96**
# feat(server): `/v1/completions` echo + logprobs + full legacy-API parity

## Why

`echo=true` + integer `logprobs` + `max_tokens=0` on legacy `/v1/completions` is how lm-eval-harness and every loglikelihood/multiple-choice evaluator scores prompts. Atlas had neither field, so likelihood-based evals (MMLU/HellaSwag/lambada-style) could not run against it — only generative ones. vLLM and llama.cpp both support this surface. A user asked exactly this ("does Atlas lack echo+logprobs?"); a full OpenAI-spec gap analysis confirmed the completions endpoint was the one place Atlas materially trailed the spec.

## What

**`/v1/completions` (the bulk):**
- `echo`, integer `logprobs` (top-k), and a true `max_tokens=0` scoring path (previously one token leaked out — the `max_tokens <= 1` guard sampled and emitted `first`). Empty output now derives `finish_reason="length"`, `completion_tokens=0`.
- **Prompt-token logprobs from prefill**: hidden states for every prompt position already existed in the prefill buffers; only the last row was ever projected. When requested, the model now norms + `lm_head_batched`-projects every position in ≤32-row batches (the logits buffer holds `min(m,32)` rows), copies each batch D2H, and records `log P(tokens[i+1] | tokens[..=i])` + top-k. Runs strictly **before** `finalize_last` (which re-derives the first sampled token's logits, so the sampler pointer stays correct). Zero cost when not requested; collecting requests bypass the prefix cache (a cache/Marconi skip would leave positions without hidden rows) and run single-stream (excluded from Q12 batched prefill).
- **Legacy response shape**: four parallel arrays `tokens` / `token_logprobs` / `top_logprobs` / `text_offset`, `null` for the first echoed token; `text_offset` = cumulative per-token decoded lengths (same approximation vLLM uses).
- **Batched prompts per spec**: `prompt: [...]` (strings or token arrays) = one independent prompt per element, one choice each, prompt-major indices. Previously arrays were joined/flattened into a single prompt — that silently corrupts lm-eval at `batch_size > 1`.
- Generated-token logprobs (previously hardcoded `None`), `n` (sequential loop, per-choice seed offset), `stream_options.include_usage` (usage-only `choices: []` chunk before `[DONE]`, chat parity), `system_fingerprint`, streaming echo (prompt chunk with logprobs precedes generation), parse-and-ignore `user`/`suffix`/`best_of`.

**Chat fixes (small):**
- `logprobs: true` alone now enables sampled-token logprobs (`top_logprobs` defaults to 0 alternatives, per spec); previously the bool was ignored.
- `developer` role (o-series successor of `system`) maps to `system` at the chat-path template conversion; previously it passed through verbatim and most templates dropped it.

**SSOT refactor:** the log-softmax + top-k math moved to `spark_model::traits::logprobs`; the scheduler's `extract_logprobs_from_f32` delegates.

## Verification

- **CI (GPU-free)**: `cargo clippy --workspace --tests` clean · `cargo fmt --check` clean · **1678 unit tests, 0 failed** (incl. new serde round-trips, legacy-shape assembly incl. null-first/text_offset/index-ordering, pure log-softmax equality, chat logprobs-resolution).
- **GPU feature matrix** (GB10, unsloth/Qwen3.6-27B-NVFP4, image `atlas-gb10:echologprobs`) — all pass:
  - echo+logprobs+max_tokens=0: 7 prompt tokens → 7 entries, first `null`, all finite ≤ 0, `finish_reason="length"`, `completion_tokens=0`, exact echo text + offsets
  - echo+generation concatenation (parallel arrays across prompt+gen), batch of 2 prompts → 2 choices with correct texts/indices/summed usage, n=2 indices, no-echo generated logprobs, chat `logprobs:true` (empty `top_logprobs`), `developer` role honored
  - streaming: echo chunk (text+logprobs, null-first) precedes tokens; finish chunk then `choices:[]` usage chunk then `[DONE]` with `include_usage`
- **lm-eval-harness acceptance** (the motivating client): `local-completions`, `lambada_openai --limit 20 --batch_size 4` → runs clean end-to-end (batched prompts exercised), **acc 0.50 ± 0.11, perplexity 4.31** — sane numbers for a chat-tuned NVFP4 27B on n=20.
- **GPU integration test** added (`#[ignore]`, `prompt_logprobs_collection_during_prefill`): prompt_len−1 entries targeting the actual next tokens, finite non-positive logprobs, descending top-k bounded by top-1, and the control (flag unset) collects nothing.
- **Flagship regression matrix** (one per DGX, same image): 35B-A3B-FP8 webserver_ok N=10 (lm-head nvfp4); 27B unsloth BFCL ST-995 subset vs the fresh same-box full-995 baseline (89.59 norm); 35B ST-995 subset old-image-vs-new-image A/B. *(results below)*

### Flagship regression results (all on `atlas-gb10:echologprobs` = this branch)

**35B-A3B-FP8 `webserver_ok` N=10 (dgx1, exact PR #278 recipe: `--lm-head-dtype nvfp4`, `ATLAS_SSM_TAIL_PROTECT=1`, cargo-shim, fixed harness):**
| Arm | Σwall | webserver_ok | followed_directions |
|---|---|---|---|
| base `main-438ed09` (same box) | 1285s | 10/10 | 10/10 |
| **this branch, run 1** | 1584s | **10/10** | **10/10** |
| **this branch, run 2** | **1388s** | **10/10** | **10/10** |

Correctness clean sweep, 0 caps/hangs in all 30 iterations. Σwall < 1500 gate met (1388); the 1285–1584 spread is temp-0.3 tier noise (single 230–320s outlier iterations swing Σ ±200 in both arms).

**35B BFCL ST-995 full-1004, old-image-vs-this-branch A/B (dgx3, seed 42, temp 0):**
both legs **84.66 overall / 83.32 normalized** — and the event logs are **token-identical** (2010/2010 content events byte-equal after stripping timestamps). The new binary reproduces the old binary's outputs exactly.

**27B unsloth BFCL ST-995 full-1004, old-image-vs-this-branch A/B (dgx2, identical config/seed):** old **87.35 / 87.54**, this branch **87.45 / 87.60** — new ≥ old on both aggregates. (~18% of events differ textually — MTP-induced greedy nondeterminism on the dense 27B, the known behavior; on the 35B, where MTP self-disables, outputs were 100% token-identical.)

## Out of scope (follow-ups)
Real `/v1/embeddings` (needs an embedding checkpoint + `embed()` path — this PR lays the per-position projection groundwork), vLLM `guided_*` aliases, `prompt_logprobs` on chat, `n` in streaming.

## Authorship
AI-generated (Claude), directed and reviewed by the maintainer. Per CONTRIBUTING §AI-First.

### CODE REVIEW

File: crates/spark-server/src/api/completions_exec.rs [hunk 1]
```diff
@@ -0,0 +1,210 @@
...
+
+/// Run every (prompt × n) choice sequentially and assemble the response.
+pub(super) async fn run_blocking(
+    state: Arc<AppState>,
+    req: &CompletionRequest,
+    prompts: Vec<Vec<u32>>,
+    p: CompletionParams,
+) -> Response {
+    let n = req.n.max(1);
+    let mut choices: Vec<CompletionChoice> = Vec::with_capacity(prompts.len() * n);
```
[the remark this answers is not in the document]
Comment (tbraun96) [reply crates/spark-server/src/api/completions_exec.rs]: Fixed in d81fb954: `n` is now validated to the OpenAI spec range (1..=128, 400 otherwise) at the handler entry, and the `Vec::with_capacity` hint is clamped independently of user input (`saturating_mul(..).min(1024)`). The same bound also caps the per-choice inference loop.
